File: frontend/lib/models/training.dart
import 'package:flutter/material.dart';

enum TrainingType {
  speed,
  control_ball,
  shoot,
  pass,
}

enum SkillLevel {
  beginner,
  intermediate,
  advanced,
  expert,
}

class TrainingSession {
  final String id;
  final String playerId;
  final TrainingType type;
  final DateTime date;
  final double score;
  final double duration;
  final SkillLevel level;
  final double reward;
  final String? videoUrl;
  final String? aiFeedback;
  final Map<String, dynamic> metrics;

  TrainingSession({
    required this.id,
    required this.playerId,
    required this.type,
    required this.date,
    required this.score,
    required this.duration,
    required this.level,
    required this.reward,
    this.videoUrl,
    this.aiFeedback,
    this.metrics = const {},
  });

  factory TrainingSession.fromJson(Map<String, dynamic> json) {
    // ✅ Gérer les valeurs null avec des valeurs par défaut
    return TrainingSession(
      id: json['id']?.toString() ?? 'unknown_${DateTime.now().millisecondsSinceEpoch}',
      playerId: json['player_id']?.toString() ?? json['playerId']?.toString() ?? 'unknown',
      type: _parseTrainingType(json['type']?.toString() ?? 'speed'),
      date: _parseDateTime(json['date'] ?? json['created_at']),
      score: (json['score'] as num?)?.toDouble() ?? 0.0,
      duration: (json['duration'] as num?)?.toDouble() ?? 0.0,
      level: _parseSkillLevel(json['level']?.toString() ?? 'beginner'),
      reward: (json['reward'] as num?)?.toDouble() ?? 0.0,
      videoUrl: json['video_url']?.toString() ?? json['videoUrl']?.toString(),
      aiFeedback: json['ai_feedback']?.toString() ?? json['feedback']?.toString(),
      metrics: json['metrics'] as Map<String, dynamic>? ?? {},
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'player_id': playerId,
      'type': type.name,
      'date': date.toIso8601String(),
      'score': score,
      'duration': duration,
      'level': level.name,
      'reward': reward,
      'video_url': videoUrl,
      'ai_feedback': aiFeedback,
      'metrics': metrics,
    };
  }

  static TrainingType _parseTrainingType(String value) {
    try {
      return TrainingType.values.firstWhere(
        (e) => e.name == value.toLowerCase(),
        orElse: () => TrainingType.speed,
      );
    } catch (e) {
      return TrainingType.speed;
    }
  }

  static SkillLevel _parseSkillLevel(String value) {
    try {
      return SkillLevel.values.firstWhere(
        (e) => e.name == value.toLowerCase(),
        orElse: () => SkillLevel.beginner,
      );
    } catch (e) {
      return SkillLevel.beginner;
    }
  }

  static DateTime _parseDateTime(dynamic value) {
    if (value == null) return DateTime.now();
    if (value is DateTime) return value;
    if (value is String) {
      try {
        return DateTime.parse(value);
      } catch (e) {
        return DateTime.now();
      }
    }
    return DateTime.now();
  }
}

File: backend/bd/models/training.py
from datetime import datetime
from enum import Enum
from sqlalchemy import String, DateTime, Float, JSON, Enum as SQLEnum, ForeignKey, Column
from sqlalchemy.orm import relationship
import uuid
from . import db

class TrainingType(Enum):
    SPEED = "speed"
    CONTROL_BALL = "control_ball"
    SHOOT = "shoot"
    PASS = "pass"

class SkillLevel(Enum):
    BEGINNER = "beginner"
    INTERMEDIATE = "intermediate"
    ADVANCED = "advanced"
    EXPERT = "expert"

class TrainingSession(db.Model):
    __tablename__ = 'training_sessions'
    
    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    player_id = Column(String(36), ForeignKey('users.id'), nullable=False, index=True)
    type = Column(SQLEnum(TrainingType), nullable=False)
    date = Column(DateTime, default=datetime.utcnow)
    score = Column(Float, nullable=False)
    duration = Column(Float, nullable=False)
    level = Column(SQLEnum(SkillLevel), nullable=False)
    reward = Column(Float, default=0.0)
    video_url = Column(String(255), nullable=True)
    ai_feedback = Column(String(500), nullable=True)
    metrics = Column(JSON, default=dict)
    created_at = Column(DateTime, default=datetime.utcnow)
    
    # Relations
    player = relationship('User', back_populates='trainings')
    
    def to_dict(self):
        return {
            'id': self.id,
            'player_id': self.player_id,
            'type': self.type.value if self.type else None,
            'date': self.date.isoformat() if self.date else None,
            'score': self.score,
            'duration': self.duration,
            'level': self.level.value if self.level else None,
            'reward': self.reward,
            'video_url': self.video_url,
            'ai_feedback': self.ai_feedback,
            'metrics': self.metrics or {},
            'created_at': self.created_at.isoformat() if self.created_at else None,
        }
    
    def __repr__(self):
        return f'<TrainingSession {self.id} - {self.type}>'